flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java

```diff
@@ -758,6 +771,76 @@ private List<StreamEdge> createChain(
         }
     }
 
+    private void setVertexParallelismsForDynamicGraphIfNecessary() {
```

Comment:
Better to add some comments to explain what this method is for, e.g. ensures the parallelism and maxParallelism of vertices in the same forward group to be the same; set the parallelism at early stage if possible, to avoid invalid partition reuse.